### ISSUE TITLE
java.nio.charset.MalformedInputException: Input length = 1 in 1.5

### ISSUE THREAD

**Issue by yazun** [tada/pljava#340]
After migrating to pljava1.6 we see quite often 
`java.nio.charset.MalformedInputException: Input length = 1` error 
even when function called do not deal with any streams in an obvious way (there might be small chance some properties are read, to be checked).
All of these functions worked ok in 1.5 - do you have any hint by any chance why this could happen? 
Were there any changes in 1.6 that could cause it?

It's not 100% reproducible and hard to pinpoint esp. we get it on postgres-xl/tbase
Setting client_encoding does not necessarily help. 

Thanks

**Comment by jcflack** [comment]
Hmm. What is the `server_encoding` on the database in question.

There were some character codec changes in 1.6 but none that (were intended to) affect situations that already worked. 1.6 is now more usable in a database with `SQL_ASCII` encoding, but that was not well supported in 1.5 so I would guess you were not using that.

Even if that were the cause, I would have expected `UnmappableCharacterException` to be more likely than `MalformedInputException`. It is possible that somewhere I have made what might be called a mistake.

Depending on where the exception is being thrown, setting the PostgreSQL logging level to `DEBUG1` or finer, in a fresh connection before the first use of any function that loads PL/Java, may cause a stack trace to be produced when the exception is thrown. The stack trace goes to the standard error of the backend process. If `logging_collector` is on, that should be collected and included in the PostgreSQL log. Otherwise, it may be system-dependent where the backend's standard error goes. It is sufficient to set either `client_min_messages` or `log_min_messages` to a `DEBUG1` or finer level.

It would be helpful to see the stack trace for a better idea where the failing conversion is happening.

There may be some sites in the code from which an exception will not generate a debug stack trace as described above. If the above procedure does not cause output of a stack trace, other techniques may be necessary.

**Comment by yazun** [comment]
Sorry for late response, it was bit difficult to get the window to test it.

`server_encoding` has been SQL_ASCII for a while, - I wanted to write _also with 1.5_, but in fact this error **is on 1.5 ** and it used to be ok before (so 1.4) (subject edited)

Finally it was possible to make it repeatable: an innocent query - without pljava calls - was needed to enable this error in the same session, otherwise everything works ok.

Stack trace is not showing up, this is all we got
```
DEBUG:  Dropped portal  (prepared statement (null)) and removed entry from the hash table
DEBUG:  Looking up portal  in the hash table
DEBUG:  Created portal  and inserted an entry in the has table
DEBUG:  find_in_dynamic_libpath: trying "/home/pgxzDR3/XZ_10_STABLE/lib/postgresql/pljava"
DEBUG:  find_in_dynamic_libpath: trying "/home/pgxzDR3/XZ_10_STABLE/lib/postgresql/pljava.so"
DEBUG:  Added JVM option string "-Xmx512M"
DEBUG:  Added JVM option string "-Dgaia.cu1.tools.util.DefaultGaiaFactory.useClassCache=false"
DEBUG:  Added JVM option string "-Dfile.encoding=UTF-8"
DEBUG:  Added JVM option string "-Dvisualvm.display.name=PL/Java:18490:surveys"
DEBUG:  Added JVM option string "vfprintf"
DEBUG:  Added JVM option string "-Xrs"
DEBUG:  Added JVM option string "-Djava.class.path=/home/pgxzDR3/XZ_10_STABLE/share/postgresql/pljava/pljava-1.5-SNAPSHOT.jar"
DEBUG:  creating Java virtual machine
DEBUG:  successfully created Java virtual machine
DEBUG:  checking for a PL/Java Backend class on the given classpath
DEBUG:  successfully loaded Backend class
DEBUG:  pgxc_abort_connections 0 coordinator to handle
DEBUG:  pgxc_abort_connections 0 datanode to handle
DEBUG:   Begin to recheck pgxc_abort_connections 0 coordinator handle
DEBUG:   Begin to recheck pgxc_abort_connections 0 datanode handle
DEBUG:  pgxc_node_remote_abort - dn_conn_count 0, co_conn_count 0
DEBUG:  pgxc_node_remote_cleanup_all - handles->co_conn_count 0,handles->dn_conn_count 0
DEBUG:  temp_object_included 0
DEBUG:  [re]setting xid = 0, old_value = 0
ERROR:  java.nio.charset.MalformedInputException: Input length = 1
```

Trying to build 1.6 now to see if it solves the problem (lso for delayed multi-dimensional array support issue) but it fails. Will create another issue for this.

**Comment by yazun** [comment]
Forgot to mention, server_encoding is indeed `SQL_ASCII` and OS LANG is `C`.

**Comment by jcflack** [comment]
Hmm, 1.5 not 1.6, that changes things. `SQL_ASCII` is a weird encoding where half of the possible bytes have known mappings to Unicode and the other half just don't, so it is difficult to deal safely with in Java where all character data must by definition be Unicode.

PL/Java before 1.5 was somewhat careless about Unicode (issue #21). 1.5 became careful, so that could definitely explain a `MalformedInputException` if there is some string value being processed that contains a non-ASCII byte. The debug log seems to show that exception coming very early, before the expected "PL/Java loaded", which should be helpful even without the stack trace. The value in question must be something encountered very early in PL/Java startup.

1.6 adds an `SQL_ASCII` "encoding" (described [here][sqla]) which should not fail for such values, so it may indeed solve the problem, though I can't say that I've tested with a value so early in initialization.

[sqla]: https://tada.github.io/pljava/use/charsets.html

**Comment by jcflack** [comment]
Believed resolved in 1.6.3 and 1.5.8.
